src/main/kotlin/com/openlattice/organizations/ExternalDatabaseManagementService.kt

```diff
@@ -0,0 +1,369 @@
+        tablesToDelete.forEach { table ->
+            organizationExternalDatabaseTables.remove(table.id)
+            aclKeyReservations.release(table.id)
+            val belongsToDeletedTable = Predicates.equal("tableId", table.id)
+            val columnsToDelete = organizationExternalDatabaseColumns.values(belongsToDeletedTable)
+            columnsToDelete.forEach { column ->
+                organizationExternalDatabaseColumns.remove(column.id)
+                aclKeyReservations.release(column.id)
+            }
+        }
+
+        //drop db from schema
+        val dbName = PostgresDatabases.buildOrganizationDatabaseName(orgId)
+        assemblerConnectionManager.connect(dbName).use {
+            it.connection.createStatement().use { stmt ->
+                stmt.execute("DROP DATABASE $dbName")
+            }
+        }
+    }
+
+    /*PERMISSIONS*/
+    fun addTrustedUser(orgId: UUID, userPrincipal: Principal, ipAdresses: Set<String>) {
+        val dbName = PostgresDatabases.buildOrganizationDatabaseName(orgId)
+        val userName = getDBUser(userPrincipal.id)
+        //edit the pg_hba file through some magic. must. become. magician.
+    }
+
+    /**
+     * Sets privileges for a user on an organization's table or column
+     */
+    fun executePrivilegesUpdate(action: Action, acls: List<Acl>) {
+        val aclsByOrg = acls.groupBy { it.aclKey[0] }
+        aclsByOrg.forEach { (orgId, acls) ->
+            val dbName = PostgresDatabases.buildOrganizationDatabaseName(orgId)
+            assemblerConnectionManager.connect(dbName).use { dataSource ->
+                dataSource.connection.autoCommit = false
+                dataSource.connection.createStatement().use { stmt ->
+                    acls.forEach {
+                        val tableAndColumnNames = getTableAndColumnNames(AclKey(it.aclKey))
+                        it.aces.forEach { ace ->
+                            val dbUser = getDBUser(ace.principal.id)
+                            val privileges = mutableListOf<String>()
+                            if (ace.permissions.contains(Permission.OWNER)) {
+                                privileges.add("ALL")
+                            } else {
+                                if (ace.permissions.contains(Permission.WRITE)) {
+                                    privileges.addAll(listOf("INSERT", "UPDATE"))
+                                }
+                                if (ace.permissions.contains(Permission.READ)) {
+                                    privileges.add("SELECT")
+                                }
+                            }
+                            val sql = createPrivilegesSql(action, privileges, tableAndColumnNames.first, tableAndColumnNames.second, dbUser)
+                            if (action == Action.SET) {
+                                val revokeSql = createPrivilegesSql(Action.REMOVE, listOf("ALL"), tableAndColumnNames.first, tableAndColumnNames.second, dbUser)
```

Comment:
Shouldn't the revoke be done before the add?

Reply:
Okay nvm, I see the revoke is done first. So this is minor, but I'd move this if statement section to the top -- I feel like it's a little more intuitive to see
```
define revokeSql
addBatch(revokeSql)
define addSql
addBatch(addSql)
```

than

```
define addSql
define revokeSql
addBatch(revokeSql)
addBatch(addSql)
```

Reply:
for sure, I'll change that